moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp

```diff
@@ -156,16 +156,77 @@ bool ChompPlanner::solve(const planning_scene::PlanningSceneConstPtr& planning_s
 
   ros::WallTime create_time = ros::WallTime::now();
 
-  ChompOptimizer optimizer(&trajectory, planning_scene, req.group_name, &params, start_state);
-  if (!optimizer.isInitialized())
+  int replan_count = 0;
+  bool replan_flag = false;
+  double org_learning_rate = 0.04, org_ridge_factor = 0.0, org_planning_time_limit = 10;
+  int org_max_iterations = 200;
+
+  // storing the initial chomp parameters values
+  org_learning_rate = params.learning_rate_;
+  org_ridge_factor = params.ridge_factor_;
+  org_planning_time_limit = params.planning_time_limit_;
+  org_max_iterations = params.max_iterations_;
+
+  ChompOptimizer* optimizer;
+
+  // create a non_const_params variable which stores the non constant version of the const params variable
+  ChompParameters params_nonconst = params_nonconst.getNonConstParams(params);
+
+  // while loop for replanning (recovery behaviour) if collision free optimized solution not found
+  while (true)
   {
-    ROS_ERROR_STREAM_NAMED("chomp_planner", "Could not initialize optimizer");
-    res.error_code.val = moveit_msgs::MoveItErrorCodes::PLANNING_FAILED;
-    return false;
-  }
-  ROS_DEBUG_NAMED("chomp_planner", "Optimization took %f sec to create", (ros::WallTime::now() - create_time).toSec());
+    if (replan_flag)
+    {
+      // increase learning rate in hope to find a successful path; increase ridge factor to avoid obstacles; add 5
+      // additional secs in hope to find a solution; increase maximum iterations
+      params_nonconst.setRecoveryParams(params_nonconst.learning_rate_ + 0.02, params_nonconst.ridge_factor_ + 0.002,
+                                        params_nonconst.planning_time_limit_ + 5, params_nonconst.max_iterations_ + 50);
+    }
```

Comment:
Shouldn't the adjustment of parameters happen *after* the first call to to ChompOptimizer?

Reply:
the `replan_flag` is `false` initially so when ChompOptimizer is first called it uses the default parameters without any adjustment supplied in the `chomp_planning.yaml`